src/mscorlib/shared/System/Globalization/DateTimeFormat.cs

```diff
@@ -1060,101 +1113,204 @@ internal static StringBuilder FormatStringBuilder(DateTime dateTime, ReadOnlySpa
 
-            dateTime.GetDatePart(out int year, out int month, out int day);
-            AppendNumber(result, year, 4);
-            result.Append('-');
-            AppendNumber(result, month, 2);
-            result.Append('-');
-            AppendNumber(result, day, 2);
-            result.Append('T');
-            AppendHHmmssTimeOfDay(result, dateTime);
-            result.Append('.');
+            uint temp = '0' + value;
+            value /= 10;
+            destination[offset + 1] = (char)(temp - (value * 10));
+            destination[offset] = (char)('0' + value);
+        }
 
-            long fraction = dateTime.Ticks % TimeSpan.TicksPerSecond;
-            AppendNumber(result, fraction, 7);
+        /// <summary>
+        /// Writes a value [ 0000 .. 9999 ] to the buffer starting at the specified offset.
+        /// This method performs best when the starting index is a constant literal.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void WriteFourDecimalDigits(uint value, Span<char> buffer, int startingIndex = 0)
+        {
+            Debug.Assert(0 <= value && value <= 9999);
 
-            FormatCustomizedRoundripTimeZone(dateTime, offset, result);
+            uint temp = '0' + value;
+            value /= 10;
+            buffer[startingIndex + 3] = (char)(temp - (value * 10));
 
-            return result;
-        }
+            temp = '0' + value;
+            value /= 10;
+            buffer[startingIndex + 2] = (char)(temp - (value * 10));
 
-        private static void AppendHHmmssTimeOfDay(StringBuilder result, DateTime dateTime)
-        {
-            // HH:mm:ss
-            AppendNumber(result, dateTime.Hour, 2);
-            result.Append(':');
-            AppendNumber(result, dateTime.Minute, 2);
-            result.Append(':');
-            AppendNumber(result, dateTime.Second, 2);
+            temp = '0' + value;
+            value /= 10;
+            buffer[startingIndex + 1] = (char)(temp - (value * 10));
+
+            buffer[startingIndex] = (char)('0' + value);
         }
 
-        internal static void AppendNumber(StringBuilder builder, long val, int digits)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void WriteDigits(ulong value, Span<char> buffer)
         {
-            for (int i = 0; i < digits; i++)
-            {
-                builder.Append('0');
-            }
+            // We can mutate the 'value' parameter since it's a copy-by-value local.
+            // It'll be used to represent the value left over after each division by 10.
 
-            int index = 1;
-            while (val > 0 && index <= digits)
+            for (int i = buffer.Length - 1; i >= 1; i--)
             {
-                builder[builder.Length - index] = (char)('0' + (val % 10));
-                val = val / 10;
-                index++;
+                ulong temp = '0' + value;
+                value /= 10;
```

Comment:
If you add the DivMod helper, we can call it in all these places.

Reply:
> If you add the DivMod helper

Math already has a DivRem method.

Reply:
Ah yes, we have access to that here.

Reply:
DivRem isn't much better (difference between int parameters vs uint parameters):
![image](https://user-images.githubusercontent.com/6527137/37729157-72b1cf1a-2cf9-11e8-9ece-377ecff47750.png)

Reply:
Also note that this code came from Utf8Formatter.